src/git.rs

```diff
@@ -80,17 +80,83 @@ pub struct Dependency {
     pub package: Option<String>,
 }
 
+pub struct RepositoryConfig {
+    pub index_location: Url,
+    pub credentials: Credentials,
+}
+
+impl RepositoryConfig {
+    pub fn from_environment() -> Self {
+        let username = dotenv::var("GIT_HTTP_USER");
+        let password = dotenv::var("GIT_HTTP_PWD");
+        let http_url = dotenv::var("GIT_REPO_URL");
+
+        let ssh_key = dotenv::var("GIT_SSH_KEY");
+        let ssh_url = dotenv::var("GIT_SSH_REPO_URL");
+
+        match (username, password, http_url, ssh_key, ssh_url) {
+            (extra_user, extra_pass, extra_http_url, Ok(encoded_key), Ok(ssh_url)) => {
+                if let (Ok(_), Ok(_), Ok(_)) = (extra_user, extra_pass, extra_http_url) {
+                    println!(
+                        "warning: both http and ssh credentials to authenticate with git are set"
+                    );
+                    println!("note: ssh credentials will take precedence over the http ones");
+                }
+
+                let index_location =
+                    Url::parse(&ssh_url).expect("failed to parse GIT_SSH_REPO_URL");
+
+                let credentials = Credentials::Ssh {
+                    key: String::from_utf8(
+                        base64::decode(&encoded_key).expect("failed to base64 decode the ssh key"),
+                    )
+                    .expect("failed to convert the ssh key to a string"),
+                };
+
+                Self {
+                    index_location,
+                    credentials,
+                }
+            }
+            (Ok(username), Ok(password), Ok(http_url), Err(_), Err(_)) => {
+                let index_location = Url::parse(&http_url).expect("failed to parse GIT_REPO_URL");
+                let credentials = Credentials::Http { username, password };
+
+                Self {
+                    index_location,
+                    credentials,
+                }
+            }
+            (_, _, Ok(http_url), _, _) => {
+                let index_location = Url::parse(&http_url).expect("failed to parse GIT_REPO_URL");
+                let credentials = Credentials::Missing;
+
+                Self {
+                    index_location,
+                    credentials,
+                }
+            }
+            _ => panic!("must have `GIT_REPO_URL` defined"),
+        }
+    }
+}
+
 pub struct Repository {
     checkout_path: TempDir,
     repository: git2::Repository,
+    credentials: Credentials,
 }
 
 impl Repository {
-    pub fn open(url: &Url, credentials: &Credentials) -> CargoResult<Self> {
+    pub fn open(repository_config: &RepositoryConfig) -> CargoResult<Self> {
         let checkout_path = TempDir::new("git")?;
+
         let repository = git2::build::RepoBuilder::new()
-            .fetch_options(Self::fetch_options(credentials))
-            .clone(url.as_str(), checkout_path.path())?;
+            .fetch_options(Self::fetch_options(&repository_config.credentials))
```

Comment:
Now that the credentials are stored in the `Repository` struct we can make `fetch_options` a method and call it like:

```rust
.fetch_options(self.fetch_options())
```

Reply:
Hm except in this spot, we don't have a `Repository` struct yet, we're in the process of creating one, and we can't create it without having the `git2::Repository`, which we can't create without the `fetch_options` here? Unless I'm misunderstanding?

Reply:
Oh, right, missed that.